src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1182 @@
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, abortException ?? ex));
+            }
+            catch (Http3ConnectionException ex)
+            {
+                // A connection-level protocol error has occurred on our stream.
+                _connection.Abort(ex);
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+            catch (Exception ex)
+            {
+                _stream.Reset((long)Http3ErrorCode.InternalError);
+                throw new IOException(SR.net_http_client_execution_error, new HttpRequestException(SR.net_http_client_execution_error, ex));
+            }
+        }
+
+        private async ValueTask<bool> ReadNextDataFrameAsync(HttpResponseMessage response, CancellationToken cancellationToken)
+        {
+            if (_dataPayloadRemaining == -1)
+            {
+                // EOS -- this branch will only be taken if user calls Read again after EOS.
+                return false;
+            }
+
+            Http3FrameType? frameType;
+            long payloadLength;
+
+            while (true)
+            {
+                (frameType, payloadLength) = await ReadFrameEnvelopeAsync(cancellationToken).ConfigureAwait(false);
+
+                switch (frameType)
+                {
+                    case Http3FrameType.Data:
+                        _dataPayloadRemaining = payloadLength;
+                        return true;
+                    case Http3FrameType.Headers:
+                        // Read any trailing headers.
+                        _trailingHeaders = new List<(HeaderDescriptor name, string value)>();
+                        await ReadHeadersAsync(payloadLength, cancellationToken).ConfigureAwait(false);
+
+                        // There may be more frames after this one, but they would all be unknown extension
+                        // frames that we are allowed to skip. Just close the stream early.
+
+                        // Note: if a server sends additional HEADERS or DATA frames at this point, it
+                        // would be a connection error -- not draining the stream means we won't catch this.
+                        goto case null;
+                    case null:
+                        // End of stream.
+                        _stream.ShutdownRead();
+                        _buffer.Dispose();
+                        CopyTrailersToResponseMessage(response);
+
+                        _dataPayloadRemaining = -1; // Set to -1 to indicate EOS.
+                        return false;
+                }
+            }
+        }
+
+        public void Trace(string message, [CallerMemberName] string memberName = null) =>
+            _connection.Trace(_stream.StreamId, message, memberName);
+
+        private sealed class Http3ReadStream : HttpBaseStream
```

Comment:
Can Http3RequestStream derive from HttpBaseStream s.t. you can avoid allocating an extra object per request?

Reply:
Possibly, but it gets weird as the users can dispose the stream. I will leave this for future TODO to look into.